src/cdomains/intDomain.ml

```diff
@@ -1825,9 +1919,468 @@ module Enums : S with type int_t = BigInt.t = struct
+  let cast_to ?torg ?(no_ov=false) t x =
+    match x with
+    | None -> None
+    | Some (c, m) when m =: Ints_t.zero ->
+      let c' = Ints_t.of_bigint @@ BigInt.cast_to t (Ints_t.to_bigint c) in
+      (* When casting into a signed type and the result does not fit, the behavior is implementation-defined. (C90 6.2.1.2, C99 and C11 6.3.1.3) *)
+      (* We go with GCC behavior here: *)
+      (*  For conversion to a type of width N, the value is reduced modulo 2^N to be within range of the type; no signal is raised. *)
+      (*   (https://gcc.gnu.org/onlinedocs/gcc/Integers-implementation.html)   *)
+      (* Clang behaves the same but they never document that anywhere *)
+      Some (c', m)
+    | _ -> match torg with
+      | (Some (Cil.TInt (ikorg, _)) ) when ikorg = t || (max_int t >= max_int ikorg && min_int t <= min_int ikorg) -> if M.tracing then M.trace "cong-cast" "some case"; x
+      | _ -> top ()
+
+
+  let cast_to ?torg ?(no_ov=false) (t : Cil.ikind) x =
+    let pretty_bool _ x = Pretty.text (string_of_bool x) in
+    let res = cast_to ?torg ~no_ov t x in
+    if M.tracing then M.trace "cong-cast" "Cast %a to %a (no_ov: %a) = %a\n" pretty x Cil.d_ikind t pretty_bool no_ov pretty res;
+    res
+
+  let widen = join
+
+  let widen ik x y =
+    let res = widen ik x y in
+    if M.tracing then M.trace "congruence" "widen %a %a -> %a\n" pretty x pretty y pretty res;
+    res
+
+  let narrow = meet
+
+  let log f ik i1 i2 =
+    match is_bot i1, is_bot i2 with
+    | true, true -> bot_of ik
+    | true, _
+    | _   , true -> raise (ArithmeticOnIntegerBot (Printf.sprintf "%s op %s" (show i1) (show i2)))
+    | _ ->
+      match to_bool i1, to_bool i2 with
+      | Some x, Some y -> of_bool ik (f x y)
+      | _              -> top_of ik
+
+  let logor = log (||)
+  let logand = log (&&)
+
+  let log1 f ik i1 =
+    if is_bot i1 then
+      bot_of ik
+    else
+      match to_bool i1 with
+      | Some x -> of_bool ik (f ik x)
+      | _      -> top_of ik
+
+  let lognot = log1 (fun _ik -> not)
+
+  let bitcomp f ik i1 i2 =
+    match is_bot i1, is_bot i2 with
+    | true, true -> bot_of ik
+    | true, _
+    | _   , true -> raise (ArithmeticOnIntegerBot (Printf.sprintf "%s op %s" (show i1) (show i2)))
+    | _ ->
+      match to_int i1, to_int i2 with
+      | Some x, Some y -> (try (of_int ik (f ik x y)) with Division_by_zero | Invalid_argument _ -> top_of ik)
+      | _              -> top_of ik
+
+  let shift_right _ _ _ = top()
+
+  let shift_right ik x y =
+    let res = shift_right ik x y in
+     if M.tracing then  M.trace "congruence" "shift_right : %a %a becomes %a \n" pretty x pretty y pretty res;
+     res
+
+  let shift_left ik x y =
+    (* Naive primality test *)
+  let is_prime n =
+    let n = Ints_t.to_int (abs n) in
+    let rec is_not_divisor d =
+      d * d > n || (n mod d <> 0 && is_not_divisor (d + 1)) in
+    n <> 1 && is_not_divisor 2
+    in
```

Comment:
Messed up indentation.

Also, why is this all done `int`? Does it not fail with large values then?

Reply:
Switched to `Ints_t` now and exposed the end recursion more clearly in the hope the compiler takes the hint

Reply:
I annotated it as such. The short circuit version is also tail-recursive, so I switched back to that.